torch/nn/modules/module.py

```diff
@@ -320,32 +323,46 @@ def register_forward_hook(self, hook):
         return handle
 
     def __call__(self, *input, **kwargs):
-        for hook in self._forward_pre_hooks.values():
-            hook(self, input)
-        result = self.forward(*input, **kwargs)
-        for hook in self._forward_hooks.values():
-            hook_result = hook(self, input, result)
-            if hook_result is not None:
-                raise RuntimeError(
-                    "forward hooks should never return any values, but '{}'"
-                    "didn't return None".format(hook))
-        if len(self._backward_hooks) > 0:
-            var = result
-            while not isinstance(var, Variable):
-                if isinstance(var, dict):
-                    var = next((v for v in var.values() if isinstance(v, Variable)))
-                else:
-                    var = var[0]
-            grad_fn = var.grad_fn
-            if grad_fn is not None:
-                for hook in self._backward_hooks.values():
-                    wrapper = functools.partial(hook, self)
-                    functools.update_wrapper(wrapper, hook)
-                    grad_fn.register_hook(wrapper)
+        tracing_state = None
+        if torch.jit._tracing:
+            input_vars = tuple(torch.autograd.function._iter_variables(input))
+            tracing_state = torch.jit.get_tracing_state(*input_vars)
+            if tracing_state:
+                tracing_state.push_scope('%s_%s' % (self.__class__.__name__, self._name))
+        try:
+            for hook in self._forward_pre_hooks.values():
+                hook(self, input)
+            result = self.forward(*input, **kwargs)
+            for hook in self._forward_hooks.values():
+                hook_result = hook(self, input, result)
+                if hook_result is not None:
+                    raise RuntimeError(
+                        "forward hooks should never return any values, but '{}'"
+                        "didn't return None".format(hook))
+            if len(self._backward_hooks) > 0:
+                var = result
+                while not isinstance(var, Variable):
+                    if isinstance(var, dict):
+                        var = next((v for v in var.values() if isinstance(v, Variable)))
+                    else:
+                        var = var[0]
+                grad_fn = var.grad_fn
+                if grad_fn is not None:
+                    for hook in self._backward_hooks.values():
+                        wrapper = functools.partial(hook, self)
+                        functools.update_wrapper(wrapper, hook)
+                        grad_fn.register_hook(wrapper)
+        finally:
+            if tracing_state:
+                tracing_state.pop_scope()
```

Comment:
I think that it would be better to do this to reduce overhead (and simplify code):
```python
# do the pre-hooks
if torch.jit._tracing:
    result = self.slow_forward() # set the scope and only then call self.forward(*input, **kwargs) in a try block
else:
    result = self.forward(*input, **kwargs)
# do the post hooks
```

Reply:
Good idea, fixed